shadowsocks/tcp.go

```diff
@@ -257,6 +278,21 @@ func drainErrToString(drainErr error) string {
 }
 
 func (s *tcpService) Stop() error {
-	s.isRunning = false
-	return s.listener.Close()
+	s.mu.Lock()
+	var err error
+	for _, l := range s.listeners {
+		if e := l.Close(); e != nil {
+			err = e
+		}
+	}
+	s.stopped = true
```

Comment:
I think only `s.stopped = true` needs to be under the Mutex, no?
But it has to be called before the Close calls.

Reply:
Any access to `s.listeners` would also need to be locked, since it could be modified by a concurrent call to `Serve`.